### PR TITLE
[bugfix] Resolve PyTorch Profiling for Manual Optimization

### LINKED ISSUE #9177: `automatic_optimization = False` breaks pytorch profiler
### To Reproduce

Run basic_examples/profiler_example.py with `--trainer.profiler pytorch` work as expected, but adding `self.automatic_optimization = False` in `ModelToProfile` breaks it with an AssertionError (even if `manual_backward` is used properly).

### Expected behavior

Profiling in manual optimization mode shall work as expected.

### Environment

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 2060
        - available:         True        
        - version:           11.1        
* Packages:
        - numpy:             1.20.1      
        - pyTorch_debug:     False       
        - pyTorch_version:   1.9.0+cu111 
        - pytorch-lightning: 1.4.4       
        - tqdm:              4.59.0      
* System:
        - OS:                Windows     
        - architecture:
                - 64bit
                - WindowsPE
        - processor:         AMD64 Family 23 Model 96 Stepping 1, AuthenticAMD
        - python:            3.8.8
        - version:           10.0.19041

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9177

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/basic_examples/profiler_example.py [hunk 1]
```diff
@@ -77,18 +94,20 @@ def train_dataloader(self, *args, **kwargs):
...
+        return torch.utils.data.DataLoader(valset, batch_size=2, shuffle=True, num_workers=0)
 
 
 def cli_main():
     if len(sys.argv) == 1:
         sys.argv += DEFAULT_CMD_LINE
 
-    LightningCLI(ModelToProfile, CIFAR10DataModule)
+    LightningCLI(
+        ModelToProfile, CIFAR10DataModule, save_config_callback=None, trainer_defaults={"profiler": PyTorchProfiler()}
```
Comment (carmocca) [inline pl_examples/basic_examples/profiler_example.py:109]:
```suggestion
        ModelToProfile, CIFAR10DataModule, save_config_overwrite=True, trainer_defaults={"profiler": PyTorchProfiler()}
```

File: pytorch_lightning/profiler/pytorch.py [hunk 2]
```diff
@@ -116,7 +116,8 @@ def pre_step(self, current_action: str) -> None:
         self._current_action = current_action
 
     def reset(self):
-        self._num_optimizer_step_and_closure = 0
+        # handle properly `fast_dev_run`. PyTorch Profiler will fail otherwise.
+        self._num_optimizer_step_and_closure = -1
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:120]: if `fast_dev_run>1`, this would be wrong?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/profiler/pytorch.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:120]: Yes. But don t have a solution so far. It seems closing the Profiler during warmup doesn t play nice.

**Review by awaelchli — APPROVED** [review]

File: tests/profiler/test_profiler.py [hunk 3]
```diff
@@ -303,50 +303,48 @@ def test_pytorch_profiler_trainer_ddp(tmpdir, pytorch_profiler):
         assert any(f"{local_rank}-validation_step" in f for f in files)
 
 
-def test_pytorch_profiler_trainer_test(tmpdir):
+@pytest.mark.parametrize("boring_model_cls", [BoringModel, ManualOptimBoringModel])
+def test_pytorch_profiler_trainer_fit(boring_model_cls, tmpdir):
```
Comment (awaelchli) [inline tests/profiler/test_profiler.py:307]: can the fit case not be covered by the test below? what could be the difference?

**Comment by awaelchli** [comment]
I copied the tests to master and they both pass. Can you check the fix / repro again?

File: pytorch_lightning/profiler/pytorch.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:120]: but why -1? are you sure it's not the sanity val check?

**Review by Borda — APPROVED** [review]

File: tests/profiler/test_profiler.py (hunk 3 above)
Comment (tchaton) [reply tests/profiler/test_profiler.py:307]: In fit, there is 2 traces generated, one for train, one for val.

Comment (tchaton) [reply tests/profiler/test_profiler.py:307]: It was more readable separated.

**Comment by tchaton** [comment]
> I copied the tests to master and they both pass. Can you check the fix / repro again?

```py
class ManualOptimBoringModel(BoringModel):
    def __init__(self):
        super().__init__()
        self.automatic_optimization = False

    def training_step(self, batch, batch_idx):
        opt = self.optimizers()
        output = self(batch)
        loss = self.loss(batch, output)
        opt.zero_grad()
        self.manual_backward(loss)
        opt.step()
        return loss

def test_pytorch_profiler_trainer_fit(tmpdir, pytorch_profiler):
    """Ensure that the profiler can be given to the training and default step are properly recorded."""
    model = ManualOptimBoringModel()
    trainer = Trainer(
        default_root_dir=tmpdir,
        progress_bar_refresh_rate=0,
        max_epochs=1,
        fast_dev_run=True,
        profiler=pytorch_profiler,
        gpus=1,
    )
    trainer.fit(model)

    expected = {"validation_step"}
    if not _KINETO_AVAILABLE:
        expected |= {"training_step_and_backward", "training_step", "backward"}
    for name in expected:
        assert sum(e.name == name for e in pytorch_profiler.function_events), name

    files = set(os.listdir(pytorch_profiler.dirpath))
    expected = f"fit-profiler-{trainer.local_rank}.txt"
    assert expected in files

    path = pytorch_profiler.dirpath / expected
    assert path.read_text("utf-8")
```

File: pytorch_lightning/profiler/pytorch.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py]: Removed !
